File: python/c29_3_polygon_area.py
import unittest
from math import fsum,isclose,ceil,floor
from itertools import pairwise
from c29_2_points_and_lines import inner_product,cross_product,point_location,point_inside_a_polygon

def shoelace_formula(p):
    return abs(fsum(cross_product(p,q) for p,q in pairwise(p)))/2

def polygon_border(p,a):
    for s1,s2 in pairwise(p):
        if isclose(point_location(s1,s2,a),0):
            if 0<=inner_product(s2-s1,a-s1)<=abs(s2-s1)**2:
                return True
    return False

def picks_theorem(p):
    a=b=0
    ymin,ymax=ceil(min(i.imag for i in p)),floor(max(i.imag for i in p))
    xmin,xmax=ceil(min(i.real for i in p)),floor(max(i.real for i in p))
    for y in range(ymin,ymax+1):
        for x in range(xmin,xmax+1):
            if polygon_border(p,complex(x,y)):
                b+=1
            elif point_inside_a_polygon(p,complex(x,y)):
                a+=1
    return a+b/2-1

class Test(unittest.TestCase):
    def test(self):
        p=[2+4j,5+5j,7+3j,4+1j,4+3j,2+4j]
        self.assertEqual(shoelace_formula(p),17/2)
        self.assertEqual(picks_theorem(p),17/2)

if __name__=='__main__':
    unittest.main()

File: c++/c29_3_polygon_area.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
#include "c29_2_points_and_lines.h"

namespace {
	long double shoelace_formula(const vector<P>& p) {
		return abs(ranges::fold_left(views::adjacent_transform<2>(p, [](P a, P b) { return cross_product(a, b); }), 0, plus())) / 2;
	}

	bool polygon_border(const vector<P>& p, P a) {
		for (auto [s1, s2] : views::pairwise(p)) {
			if (abs(point_location(s1, s2, a)) < 1e-9l) {
				auto i = inner_product(s2 - s1, a - s1);
				if (0 <= i && i <= inner_product(s2 - s1, s2 - s1)) {
					return true;
				}
			}
		}
		return false;
	}

	long double picks_theorem(const vector<P>& p) {
		ll a = 0, b = 0;
		auto ymin = numeric_limits<long double>::infinity(), ymax = -numeric_limits<long double>::infinity();
		auto xmin = numeric_limits<long double>::infinity(), xmax = -numeric_limits<long double>::infinity();
		for (P i : p) {
			ymin = min(ymin, i.imag());
			ymax = max(ymax, i.imag());
			xmin = min(xmin, i.real());
			xmax = max(xmax, i.real());
		}
		for (auto y = ymin; y <= ymax; ++y) {
			for (auto x = xmin; x <= xmax; ++x) {
				if (polygon_border(p, { x, y }))
					++b;
				else if (point_inside_a_polygon(p, { x, y }))
					++a;
			}
		}
		return a + b / 2. - 1;
	}
}

TEST(C293PolygonArea, shoelaceFormula) {
	EXPECT_EQ(shoelace_formula({ {2,4},{5,5},{7,3},{4,1},{4,3},{2,4} }), 17 / 2.);
}

TEST(C293PolygonArea, picksTheorem) {
	EXPECT_EQ(picks_theorem({ {2,4},{5,5},{7,3},{4,1},{4,3},{2,4} }), 17 / 2.);
}
